TanStack Start SDK (#1495)
clerk/clerk-docs: docs/quickstarts/tanstack-start.mdx

---
title: Tanstack Start Quickstart (Beta)
description: Learn how to use Clerk to quickly and easily add secure authentication and user management to your TanStack Start application.
---

> [!WARNING]
> The TanStack Start SDK is currently in beta. **It is not yet recommended for production use**.

<TutorialHero
  framework="react"
  exampleRepo={[
    {
      title: "TanStack Start Quickstart Repo",
      link: "https://github.com/clerk/clerk-tanstack-start-quickstart"

    }
  ]}
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "/docs/quickstarts/setup-clerk",
      icon: "clerk",

>     }

Reviewer: We have a `tanstack` icon 👍

>     }

Reviewer 2: updated the comment, let's make sure that gets added @octoper :)

>     }

Author: We don't want this to be `tanstack-start` as we will use the icon for TanStack Router too and all share the same logo.

  - Set up your environment variables
  - Configure `clerkHandler`
  - Add `<ClerkProvider />` to your application
  - Protect your pages
</TutorialHero>

Learn how to use Clerk to quickly and easily add secure authentication and user management to your TanStack Start application.

<Steps>
  ### Install `@clerk/tanstack-start`

  Once you have a TanStack Start application ready, you need to install Clerk's TanStack Start SDK. This gives you access to our prebuilt components and hooks for TanStack Start applications.

  <CodeBlockTabs type="installer" options={["npm", "yarn", "pnpm"]}>
    ```bash {{ filename: 'terminal' }}
    npm install @clerk/tanstack-start
    ```

    ```bash {{ filename: 'terminal' }}
    yarn add @clerk/tanstack-start
    ```

    ```bash {{ filename: 'terminal' }}
    pnpm add @clerk/tanstack-start
    ```
  </CodeBlockTabs>

  ### Set your environment variables

  Add the following keys to your `.env` file. These keys can always be retrieved from the [API Keys](https://dashboard.clerk.com/last-active?path=api-keys) page of your Clerk Dashboard.

  ```env {{ filename: '.env' }}
  VITE_CLERK_PUBLISHABLE_KEY={{pub_key}}
  CLERK_SECRET_KEY={{secret}}

>   ```

Author: As TanStack uses Vite this should work

  ### Add `clerkHandler` to your application

>   To configure Clerk in your TanStack application, you will need to update the SSR entrypoint.

Reviewer 2: ```suggestion
  Tanstack's [`createStartHandler()`](https://tanstack.com/router/latest/docs/framework/react/start/getting-started#the-server-entry-point) creates a server-side handler that determines which routes and loaders need to be executed when the user hits a given route.

  {/* TODO: There is no reference for createClerkHandler()... what does it do */}

  Clerk's `createClerkHandler()`
  
  To configure Clerk in your TanStack application, update the SSR entrypoint by wrapping `createStartHandler()` in `createClerkHandler()`.
```

Reviewer 2: clerkHandler isn't documented ?

Author: Added!

>   ```tsx {{ filename: 'app/ssr.tsx', ins: [4, [11, 14]], del: [[6, 9]] }}
>   import { createStartHandler, defaultStreamHandler } from '@tanstack/start/server'
>   import { getRouterManifest } from '@tanstack/start/router-manifest'
>   import { createRouter } from './router'
>   import { createClerkHandler } from '@clerk/tanstack-start/server'
> 
>   export default createStartHandler({
>     createRouter,
>     getRouterManifest,
>   })(defaultStreamHandler)
> 
>   export default createClerkHandler(
>     createStartHandler({
>       createRouter,
>       getRouterManifest,
>     }),
>   )(defaultStreamHandler)
>   ```

Reviewer 2: ```suggestion
  ```tsx {{ filename: 'app/ssr.tsx' }}
  import { createStartHandler, defaultStreamHandler } from '@tanstack/start/server'
  import { getRouterManifest } from '@tanstack/start/router-manifest'
  import { createRouter } from './router'
  import { createClerkHandler } from '@clerk/tanstack-start/server'

  export default createClerkHandler(
    createStartHandler({
      createRouter,
      getRouterManifest,
    }),
  )(defaultStreamHandler)
  ```
```

  ### Add `<ClerkProvider>` and components to your app

  All Clerk hooks and components must be children of the [`<ClerkProvider>`](/docs/components/clerk-provider) component, which provides active session and user context.

>   ```tsx {{ filename: 'app/routes/__root.tsx', ins: [11, 28, 32] }}

Reviewer 2: Did you mean to highlight something here? The highlighting is way off.

```suggestion
  ```tsx {{ filename: 'app/routes/__root.tsx', }}
```

Author: This probably happened after we enabled prettier formatting for code blocks, I'm gonna fix the highlighting

  import { Outlet, ScrollRestoration, createRootRoute } from '@tanstack/react-router'
  import { TanStackRouterDevtools } from '@tanstack/router-devtools'
  import { Body, Head, Html, Meta, Scripts } from '@tanstack/start'
  import * as React from 'react'
  import { DefaultCatchBoundary } from '~/components/DefaultCatchBoundary'
  import { NotFound } from '~/components/NotFound'
  import { ClerkProvider } from '@clerk/tanstack-start'

>   import appCss from '~/styles/app.css?url'

Reviewer 2: There are a lot of imports here, and things going on in this code block that are not from the tanstack quickstart - if we are going to add things like this, we need to explain them in the doc. But these don't seem necessary for a "quickstart". Can you provide more info here on what's being added and why?

Author: Yeah we can remove some of them, most of those are taken from the early example of TanStack Start as there was no docs when we started creating the SDK

  export const Route = createRootRoute({
    links: () => [{ rel: 'stylesheet', href: appCss }],
    errorComponent: (props) => {
      return (
        <RootDocument>
          <DefaultCatchBoundary {...props} />
        </RootDocument>
      )
    },
    notFoundComponent: () => <NotFound />,
    component: () => {
      return (
        <ClerkProvider>
          <RootDocument>
            <Outlet />
          </RootDocument>
        </ClerkProvider>
      )
    },
  })

  function RootDocument({ children }: { children: React.ReactNode }) {
    return (
      <Html>
        <Head>
          <Meta />
        </Head>
        <Body>
          {children}
          <ScrollRestoration />
          <TanStackRouterDevtools position="bottom-right" />
          <Scripts />
        </Body>
      </Html>
    )
  }
  ```

  ### Protect your pages

  #### Client side

>   Clerk offers [Control Components](/docs/components/overview) that allow you to protect your pages. These components are used to control the visibility of your pages based on the user's authentication state.

Reviewer 2: ```suggestion
  To protect your pages on the client side, you can use Clerk's [prebuilt control components](/docs/components/overview#what-are-control-components) that control the visibility of content based on the user's authentication state.

  The following example uses the following components:

  - [`<SignedIn>`](/docs/components/control/signed-in): Children of this component can only be seen while **signed in**.
  - [`<SignedOut>`](/docs/components/control/signed-out): Children of this component can only be seen while **signed out**.
  - [`<UserButton />`](/docs/components/user/user-button): A prebuilt component that comes styled out of the box to show the avatar from the account the user is signed in with.
  - [`<SignInButton />`](/docs/components/unstyled/sign-in-button): An unstyled component that links to the sign-in page. For this example, it links to the [Account Portal sign-in page.](/docs/customization/account-portal/overview)
  - [`<SignUpButton />`](/docs/components/unstyled/sign-up-button): An unstyled component that links to the sign-up page or displays the sign-up modal, depending on which `mode` you have set. For this example, it links to the [Account Portal sign-up page.](/docs/customization/account-portal/overview)
```

  ```tsx {{ filename: 'app/routes/index.tsx' }}
  import {
    SignedIn,
    UserButton,
    SignOutButton,
    SignedOut,
    SignInButton,
    SignUpButton,
  } from '@clerk/tanstack-start'
  import { createFileRoute } from '@tanstack/react-router'

  export const Route = createFileRoute('/')({
    component: Home,
  })

  function Home() {
    return (
      <div>
        <h1>Index Route</h1>

>         <SignedIn>
>           <p>You are signed in!</p>
>           <div>
>             <p>View your profile here</p>
>             <UserButton />
>           </div>
>           <div>
>             <SignOutButton />
>           </div>
>         </SignedIn>

Reviewer 2: ```suggestion
        <SignedIn>
        <p>You are signed in!</p>

        <UserButton />

        <SignOutButton />
      </SignedIn>
```

        <SignedOut>
          <p>You are signed out</p>

          <SignInButton mode="modal" />

          <SignUpButton mode="modal" />
        </SignedOut>
      </div>
    )
  }
  ```

>   #### Server side

Reviewer 2: Getting an error here for `to` saying that it only accepts `/` `.` and `..` as its value ?

<img width="1157" alt="Screenshot 2024-09-06 at 13 33 08" src="https://github.com/user-attachments/assets/7dc5a7e8-d970-4438-9bc8-f1ca2ec1cfcc">

Author: This is expected as the route does not exists and everything in TanStack Start is fully typed, this is there as placeholder

      link: "https://github.com/clerk/clerk-tanstack-start-quickstart"

    }
  ]}
  beforeYouStart={[
    {
      title: "Set up a Clerk application",
      link: "/docs/quickstarts/setup-clerk",
      icon: "clerk",
    }
  ]}
>
  - Install `@clerk/tanstack-start`
  - Set up your environment variables
  - Configure clerkHandler
  - Add `<ClerkProvider/>` to your application
  - Protecting your pages
</TutorialHero>

Learn how to use Clerk to quickly and easily add secure authentication and user management to your TanStack Start application.

<Steps>
  ### Install `@clerk/tanstack-start`

  Once you have a TanStack Start application ready, you need to install Clerk's TanStack Start SDK. This gives you access to our prebuilt components and hooks for TanStack Start applications.

  <CodeBlockTabs type="installer" options={["npm", "yarn", "pnpm"]}>
    ```bash {{ prettier: false, filename: 'terminal' }}
    npm install @clerk/tanstack-start
    ```

    ```bash {{ prettier: false, filename: 'terminal' }}
    yarn add @clerk/tanstack-start
    ```

    ```bash {{ prettier: false, filename: 'terminal' }}
    pnpm add @clerk/tanstack-start
    ```
  </CodeBlockTabs>

>   ### Set environment variables

Reviewer 3: ```suggestion
  ### Set your environment variables
```

>   Below is an example of an `.env` file.
> 
>   **Pro tip!** If you are signed into your Clerk Dashboard, your secret key should become visible by clicking on the eye icon. Otherwise, you can find your keys in the Clerk Dashboard on the [API Keys](https://dashboard.clerk.com/last-active?path=api-keys) page.

Reviewer 3: ```suggestion
Add the following keys to your `.env` file. These keys can always be retrieved from the [API Keys](https://dashboard.clerk.com/last-active?path=api-keys) page of your Clerk Dashboard.
```

Aligns with what all the other quickstarts do.

  ```env {{ prettier: false, filename: '.env' }}
  VITE_CLERK_PUBLISHABLE_KEY={{pub_key}}
  CLERK_SECRET_KEY={{secret}}
  ```

  ### Add `clerkHandler` to your application

>   To configure Clerk in your TanStack application, you will need to update SSR entrypoint.

Reviewer 3: ```suggestion
  To configure Clerk in your TanStack application, you will need to update the SSR entrypoint.
```

>   ```tsx {{ prettier: false, filename: 'app/ssr.tsx' }}

Reviewer 3: How does the default `app/ssr.tsx` file look like? Comparing it to that we can use line highlighting to outline the changes better

>   ```tsx {{ prettier: false, filename: 'app/ssr.tsx' }}

Reviewer 3: ```suggestion
  ```tsx {{ filename: 'app/ssr.tsx' }}
```

Disabling prettier should be an escape hatch, let's remove it from the tsx codeblocks!

  import { getRouterManifest } from '@tanstack/start/router-manifest';
  import { createRouter } from './router';
  import { createClerkHandler } from '@clerk/tanstack-start/server';

  export default createClerkHandler(createStartHandler({
    createRouter,
    getRouterManifest,
  }))(defaultStreamHandler);
  ```

  ### Add `<ClerkProvider>` and components to your app

  All Clerk hooks and components must be children of the [`<ClerkProvider>`](/docs/components/clerk-provider) component, which provides active session and user context.

  You can control which content signed in and signed out users can see with Clerk's [prebuilt components](/docs/components/overview). To get started, create a header for your users to sign in or out. To do this, you will use:

  - [`<SignedIn>`](/docs/components/control/signed-in): Children of this component can only be seen while **signed in**.
  - [`<SignedOut>`](/docs/components/control/signed-out): Children of this component can only be seen while **signed out**.
  - [`<UserButton />`](/docs/components/user/user-button): A prebuilt component that comes styled out of the box to show the avatar from the account the user is signed in with.

>   - [`<SignInButton />`](/docs/components/unstyled/sign-in-button): An unstyled component that links to the sign-in page or displays the sign-in modal.

Reviewer 3: The example below doesn't do that, it only adds the Provider 😊 So we either adjust the example or this text

>   Select your preferred router to learn how to make this data available across your entire app:

Reviewer 3: I know this was copy/pasted from Next.js, since you won't choose between routers here this needs changing

>   ```tsx {{ prettier: false, filename: 'app/routes/__root.tsx' }}

Reviewer 3: ```suggestion
  ```tsx {{ filename: 'app/routes/__root.tsx' }}
```

    import * as React from 'react'
    import { DefaultCatchBoundary } from '~/components/DefaultCatchBoundary'
    import { NotFound } from '~/components/NotFound'
    import { ClerkProvider } from '@clerk/tanstack-start'
    import appCss from '~/styles/app.css?url'

    export const Route = createRootRoute({
      links: () => [
        { rel: 'stylesheet', href: appCss },
      ],
      errorComponent: (props) => {
        return (
          <RootDocument>
            <DefaultCatchBoundary {...props} />
          </RootDocument>
        )
      },
      notFoundComponent: () => <NotFound />,
      component: () => {
        return (
          <ClerkProvider>
            <RootDocument>
              <Outlet />
            </RootDocument>
          </ClerkProvider>
        )
      },
    })

    function RootDocument({ children }: { children: React.ReactNode }) {
      return (
        <Html>
          <Head>
            <Meta />
          </Head>
          <Body>
            {children}
            <ScrollRestoration />
            <TanStackRouterDevtools position="bottom-right" />
            <Scripts />
          </Body>
        </Html>
      )
    }
  ```

>   ### Protecting your pages

Reviewer 3: ```suggestion
  ### Protect your pages
```

  #### Client side

  Clerk offers [Control Components](/docs/components/overview) that allow you to protect your pages. These components are used to control the visibility of your pages based on the user's authentication state.

>   ```tsx {{ prettier: false, filename: 'app/routes/index.tsx' }}

Reviewer 3: ```suggestion
  ```tsx {{ filename: 'app/routes/index.tsx' }}
```

  import { SignedIn, UserButton, SignOutButton, SignedOut, SignInButton, SignUpButton } from '@clerk/tanstack-start'
  import { createFileRoute } from '@tanstack/react-router'

  export const Route = createFileRoute('/')({
    component: Home,
  })

  function Home() {
    return (
      <div>
        <h1>Index Route</h1>
        <SignedIn>
          <p>You are signed in!</p>
          <div>
            <p>View your profile here</p>
            <UserButton />
          </div>
          <div>
            <SignOutButton />
          </div>
        </SignedIn>
        <SignedOut>
          <p>You are signed out</p>

          <SignInButton mode='modal' />

          <SignUpButton mode='modal' />
        </SignedOut>
      </div>
    )
  }
  ```

  #### Server side

  To protect your routes, you can use the `beforeLoader` paired with server function to check for the `userId` singleton. If it doesn't exist, redirect your user back to the sign-in page.

>   ```tsx {{ prettier: false, filename: 'app/routes/index.tsx' }}

Reviewer 3: ```suggestion
  ```tsx {{ filename: 'app/routes/index.tsx' }}
```

  - [`<SignInButton />`](/docs/components/unstyled/sign-in-button): An unstyled component that links to the sign-in page. For this example, it links to the [Account Portal sign-in page.](/docs/customization/account-portal/overview)
  - [`<SignUpButton />`](/docs/components/unstyled/sign-up-button): An unstyled component that links to the sign-up page or displays the sign-up modal, depending on which `mode` you have set. For this example, it links to the [Account Portal sign-up page.](/docs/customization/account-portal/overview)

  ```tsx {{ filename: 'app/routes/index.tsx' }}
  import {
    SignedIn,
    UserButton,
    SignOutButton,
    SignedOut,
    SignInButton,
    SignUpButton,
  } from '@clerk/tanstack-start'
  import { createFileRoute } from '@tanstack/react-router'

  export const Route = createFileRoute('/')({
    component: Home,
  })

  function Home() {
    return (
      <div>
        <h1>Index Route</h1>
        <SignedIn>
        <p>You are signed in!</p>

        <UserButton />

        <SignOutButton />
      </SignedIn>
        <SignedOut>
          <p>You are signed out</p>

>           <SignInButton mode="modal" />
> 
>           <SignUpButton mode="modal" />

Reviewer 2: ```suggestion
          <SignInButton />

          <SignUpButton />
```
